tests/adapter/test_tiered.py

```diff
@@ -86,3 +87,60 @@ async def test_tiered(DatastoreTests, Adapter, DictDatastore, encode_fn):
 	
 	# Datastores should have been cleaned up by aexit
 	assert len(ts._stores) == 0
+
+
+@pytest.mark.parametrize(*make_datastore_test_params("tiered"))
+@trio.testing.trio_test
+async def test_tiered_half_fail(DatastoreTests, Adapter, DictDatastore, encode_fn):
+	s1 = DictDatastore()
+	s2 = DictDatastore()
+	async with Adapter([s1, s2]) as ts:
+		k1 = datastore.Key("/a")
+		k2 = datastore.Key("/b")
+		
+		assert not await s1.contains(k1)
+		assert not await s1.contains(k2)
+		assert not await s2.contains(k1)
+		assert not await s2.contains(k2)
+		
+		# Write value into only one datastore
+		await s1.put(k1, encode_fn("1"))
+		await s2.put(k2, encode_fn("1"))
+		
+		assert await s1.contains(k1)
+		assert not await s1.contains(k2)
+		assert not await s2.contains(k1)
+		assert await s2.contains(k2)
+		
+		# Write non-replacingly to both
+		with conftest.raises(KeyError):
+			await ts.put(k1, encode_fn("2"), replace=False)
+		with conftest.raises(KeyError):
+			await ts.put(k2, encode_fn("2"), replace=False)
+		
+		assert await s1.contains(k1)
+		assert not await s1.contains(k2)
+		assert not await s2.contains(k1)
+		assert await s2.contains(k2)
+		
+		# Ensure red back value didn't change
+		assert await ts.get_all(k1) == encode_fn("1")
+		assert await ts.get_all(k2) == encode_fn("1")
+		
+		assert await s1.contains(k1)
+		assert await s1.contains(k2)
```

Comment:
I may be wrong (coz the tests are passing), but shouldn't this be `assert await not s1.contains(k2)`?

Reply:
Just read through it again and can confirm that this behaviour is intended: The idea behind the tiered datastore is to accelerate access to commonly used data – it does this by storing data read from a lower store (in practice this could be a slower disk or a remote database) in all the datastores above (like fast in-memory or on-disk caches) to accelerate access to it on subsequent reads. Hence the behaviour tested here is correct for this particular datastore (it would have been different had we used `.get_all` on the underlying datastores directly).